src/inmanta/ast/statements/__init__.py

```diff
@@ -217,12 +217,28 @@ def requires(self) -> list[str]:
 
     def execute_direct(self, requires: abc.Mapping[str, object]) -> object:
         """
-        Execute this statement in a static context without any scheduling, returning the expression's result.
+        Execute this expression in a static context without any scheduling, returning the expression's result.
+
+        Used by typedefs to force immediate evaluation.
 
         :param requires: A dictionary mapping names to values.
         """
         raise DirectExecuteException(self, f"The statement {str(self)} can not be executed in this context")
 
+    def as_constant(self) -> object:
+        """
+        Returns this expression as a constant value, if possible. Otherwise, raise a RuntimeException.
+
+        It is very similar to self.execute_direct({}) with the difference that:
+         - This method can be called prior to normalization
+         - It is expected to not call plugins
```

Comment:
```suggestion
         - It is expected to not have side effects (e.g. call plugins, construct entities, ....)
```
I'm a bit on the fence about the second example. I think it's important to keep in mind if we ever decide to extend the functionality. However, it might be interpreted to mean that `execute_direct` does allow constructing entities, which it doesn't (it's surface is a bit less well defined).

Interestingly, the fact that it executes without a context in effect prohibits it from doing any of these things, even by accident. But that's by happenstance more than by design.